server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java

```diff
@@ -0,0 +1,404 @@
+
+    try {
+      var path = ZooUtil.getRoot(context.getInstanceID());
+      if (zrw.exists(path, propStoreWatcher)) {
+        log.debug("have a ZooKeeper connection");
+        zkReadyMon.setReady();
+      } else {
+        throw new IllegalStateException(
+            "Instance may not have been initialized, root node" + path + " does not exist");
+      }
+    } catch (InterruptedException | KeeperException ex) {
+      throw new IllegalStateException("Failed to read root node from ZooKeeper", ex);
+    }
+  }
+
+  /**
+   * Create initial system props for the instance. If the node already exists, no action is
+   * performed.
+   *
+   * @param context
+   *          the server context.
+   * @param initProps
+   *          map of k, v pairs of initial properties.
+   * @return true if props create, false if the node existed and initialization was skipped.
+   */
+  public synchronized static boolean initSysProps(final ServerContext context,
+      final Map<String,String> initProps) {
+    PropCacheId sysPropsId = PropCacheId.forSystem(context.getInstanceID());
+    return createInitialProps(context, sysPropsId, initProps);
+  }
+
+  /**
+   * Create initial properties if they do not exist. If the node exists, initialization will be
+   * skipped.
+   *
+   * @param context
+   *          the system context
+   * @param propCacheId
+   *          a prop id
+   * @param props
+   *          initial properties
+   * @return true if props create, false if the node existed and initialization was skipped.
+   */
+  public static boolean createInitialProps(final ServerContext context,
+      final PropCacheId propCacheId, Map<String,String> props) {
+
+    try {
+      ZooReaderWriter zrw = context.getZooReaderWriter();
+      if (zrw.exists(propCacheId.getPath())) {
+        return false;
+      }
+      VersionedProperties vProps = new VersionedProperties(props);
+      return zrw.putPersistentData(propCacheId.getPath(),
+          context.getVersionedPropertiesCodec().toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException("Interrupted creating node " + propCacheId, ex);
+    } catch (Exception ex) {
+      throw new IllegalStateException("Failed to create node " + propCacheId, ex);
+    }
+  }
+
+  public PropStoreMetrics getMetrics() {
+    return cacheMetrics;
+  }
+
+  // TODO - evaluate returning the props instead of boolean.
+  @Override
+  public boolean create(PropCacheId propCacheId, Map<String,String> props)
+      throws PropStoreException {
+
+    VersionedProperties vProps = new VersionedProperties(props);
+
+    try {
+
+      var path = propCacheId.getPath();
+      if (!zrw.putPersistentData(path, propCodec.toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL)) {
+        return false;
+      }
+
+      Stat stat = zrw.getStatus(path, propStoreWatcher);
```

Comment:
There may be a race condition here.  Something could run between the putPersistentData call and getStatus call that changes the stat to make it differ.

Reply:
I agree - but not sure what to do.  The creation of the node and then adding a watcher seems to require two ZooKeeper calls (or I'm missing something) There seems to be a few options:

1. Ignore the check. the getStatus call is only to set the watcher. It may not matter if it changed - the stored properties are only in ZooKeeper and not accessible until read and placed into the cache. That read would get whatever version is current.
2. Add a method in ZooReaderWriter that returns the Stat node from the ZooKeeper create call - and then check using that value.

Not sure if you have thoughts on which would be preferable.  This could be something that changes with using the ZooKeeper version instead of calculating it in VersionedProperties.